File: d_good_substrings.cpp
#include <bits/stdc++.h>
using namespace std;

struct TrieNode {
        TrieNode* children[26];
        bool is_end;
        int count;

        TrieNode() {
            for (int i = 0; i < 26; i++) children[i] = nullptr;
            is_end = false;
            count = 0;
        }
};

class Trie {
    public:
        TrieNode* root;
        vector<int> goods;
        long long ans;

        Trie(const vector<int>& goods) {
            this->root = new TrieNode();
            this->goods = goods;
            this->ans = 0;
        }

        void insert(const string& word, int start, int k) {
            TrieNode* node = root;
            int bad = 0;
            long long local_ans = 0;

            for (int i = start; i < (int)word.size(); i++) {
                int idx = word[i] - 'a';
                bad += (goods[idx] == 0);

                if (!node->children[idx]) {
                    node->children[idx] = new TrieNode();
                    if (bad <= k) local_ans++;
                }

                node->count = bad;
                node = node->children[idx];

                if (bad > k) {
                    ans += local_ans;
                    return;
                }
            }

            node->is_end = true;
            ans += local_ans;
        }
};

void solve() {
    string s, g;
    int k;
    cin >> s >> g >> k;

    vector<int> goods(26);
    for (int i = 0; i < 26; i++) {
        goods[i] = g[i] - '0';
    }

    Trie trie(goods);

    for (int i = 0; i < (int)s.size(); i++) {
        trie.insert(s, i, k);
    }

    cout << trie.ans << "\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    solve();
    return 0;
}

File: d_good_substrings.py
import sys

# sys.setrecursionlimit(10**6)
from collections import Counter, defaultdict, deque
from bisect import bisect_left, bisect_right
from copy import deepcopy
import math

# input = input
input = sys.stdin.readline


# ---------- Trie Implementation (Lazy Deletion with Count) ----------
class TrieNode:
    def __init__(self):
        self.children = [None] * 26
        self.is_end = False
        self.count = 0


class Trie:
    def __init__(self, goods):
        self.root = TrieNode()
        self.goods = goods
        self.ans = 0

    def _char_to_index(self, ch):
        return ord(ch) - ord("a")

    def insert(self, word, l, k):
        curr = 0
        node = self.root
        ans = 0
        for i in range(l, len(word)):
            ch = word[i]
            idx = self._char_to_index(ch)
            curr += not self.goods[idx]
            if not node.children[idx]:
                node.children[idx] = TrieNode()
                ans += curr <= k
            node.count = curr
            node = node.children[idx]
            if curr > k:
                self.ans += ans
                return
        node.is_end = True
        self.ans += ans

    def query(self, word):
        node = self.root
        for ch in word:
            idx = self._char_to_index(ch)
            child = node.children[idx]
            if not child or child.count == 0:
                return False
            node = child
        return node.is_end

    def delete(self, word):
        node = self.root
        path = []

        for ch in word:
            idx = self._char_to_index(ch)
            child = node.children[idx]
            if not child or child.count == 0:
                return  # word not found
            path.append(child)
            node = child

        if not node.is_end:
            return

        node.is_end = False
        for child in path:
            child.count -= 1


def solve():
    chars = list(x for x in input().strip())
    goods = list(int(x) for x in input().strip())
    k = int(input())
    trie = Trie(goods)
    for i in range(len(chars)):
        trie.insert(chars, i, k)
    print(trie.ans)


def main():
    t = 1
    for _ in range(t):
        solve()


if __name__ == "__main__":
    main()
